DRAFT - mdx for Synthetic Data Creation lung images (#105)
johko/computer-vision-course: chapters/en/Unit 10 - Synthetic Data Creation/synthetic-lung-images.mdx

```python
from datasets import load_dataset

dataset = load_dataset("hf-vision/chest-xray-pneumonia")
```

We will preprocess the lung images by resizing and normalizing the pixel values.

```python
import torchvision.transforms as transforms
from torchvision.transforms import CenterCrop, Compose, Normalize, Resize, ToTensor

transform=Compose(
    [
          transforms.Resize(image_size),
          transforms.CenterCrop(image_size),
          transforms.ToTensor(),
    ]
)
```

During training, the generator aims to produce synthetic lung images that are indistinguishable from real images, while the discriminator learns to correctly classify the images as real or synthetic. We start by initializing the generator with random noise and will train for 100 epochs.

Let's visualize the progress:

![trainig-gif](https://huggingface.co/datasets/hf-vision/course-assets/blob/main/dcgan-lung_images/dcgan_training_animation.gif)

## How did we do?

Here are 64 "verified" synthetic images, defined as receiving a "real" label from the discriminator with a 70% probability.

![lung-images](https://huggingface.co/datasets/hf-vision/course-assets/blob/main/dcgan-lung_images/verified.png)

> We can see that most of the synthetic lung images look good, but a couple are hazy. There are some important things to mention. First, researchers generating synthetic medical images typically employee a "human in the middle"--in this case, a expert radiologist--to evaluate the synthetic images. Then, only the images that fool the expert are included with the real data to train the model. Second, even though the generated images appear to look OK (at least to an amateur), they do appear to be very similar. This is another known issue with GANS - they suffer from "mode collapse." Essentially, this happens when the generator starts producing the same type of output. Think of it as someone who has gotten a lot of praise for making chocolate chip cookies and therefore gets _really, really_ good at making those cookies but can't make other types of cookies.

Reviewer: ```suggestion
We can see that most of the synthetic lung images look good, but a couple are hazy. There are some important things to mention. First, researchers generating synthetic medical images typically employ a "human in the middle"--in this case, an expert radiologist--to evaluate the synthetic images. Then, only the images that fool the expert are included with the real data to train the model. Second, even though the generated images appear to look OK (at least to an amateur), they do appear to be very similar. This is another known issue with GANs - they suffer from "mode collapse." Essentially, this happens when the generator starts producing the same type of output. Think of it as someone who has gotten a lot of praise for making chocolate chip cookies and, therefore, gets _really, really_ good at making those cookies but can't make any other type of cookies.
```

> [[Link to Notebook]]

Reviewer 2: You can put the link here (internally, e.g. (./name-of-notebook.ipynb)

Author: Is it just the name and not the relative path? (That is what I have added.)

Reviewer 2: relative path itself yep

is challenged with small datasets, privacy concerns, and a limited amount of annotated samples. Researchers have used GANs to generate synthetic images such as lung X-ray images, retina images,
brain scans, and liver images. In [GAN-based synthetic brain PET image generation](https://braininformatics.springeropen.com/counter/pdf/10.1186/s40708-020-00104-2.pdf), the authors
created brain PET images for three different stages of Alzheimer’s disease. [GAN-based Synthetic Medical Image Augmentation for increased CNN Performance in Liver Lesion Classification](https://arxiv.org/abs/1803.01229)
generated synthetic liver images. [BrainGAN: Brain MRI Image Generation and Classification Framework Using GAN Architectures and CNN Models](https://www.mdpi.com/1424-8220/22/11/4297) developed a
framework for generating brain MRI images using multiple GAN architectures, and [A Novel COVID-19 Detection Model Based on DCGAN and Deep Transfer Learning](https://www.sciencedirect.com/science/article/pii/S1877050922007463) used DCGAN to generate synthetic lung X-ray images to aid in COVID-19 detection.

## DCGAN (Deep Convolutional Generative Adversarial Network):

> DCGAN was proposed in [Unsupervised Representation Learning With Deep Convolutional Generative Adversarial Networks](https://arxiv.org/abs/1511.06434) by Radford, et al. and is the model

Reviewer 2: ```suggestion
DCGAN was proposed in [Unsupervised Representation Learning With Deep Convolutional Generative Adversarial Networks](https://arxiv.org/abs/1511.06434) by Radford, et al. and is the
```

> most often used for medical imaging. We are going to use it to generate synthetic lung

Reviewer 2: ```suggestion
most often used model for medical imaging. We are going to use it to generate synthetic lung
```
Can you remove the newlines?

> most often used for medical imaging. We are going to use it to generate synthetic lung

Reviewer 3: Coming from an academic background: Do we have proof that it is the most widely used model for that or is that just an educated assumption?
If there is some source that states it, it would be nice to include it.

Author: I appreciate the comment since I also have an academic background. The truth is I don't remember if I read that or that was just my impression after reading a bunch of studies, so I have edited the sentence.

images.

Here are the basic steps:

> _Data Collection:_

Reviewer 2: ```suggestion
**Data Collection**
```
We can make smallest headers with bold

> First, We need to obtain a dataset of real lung images.

Reviewer 2: ```suggestion
First, we need to obtain a dataset of real lung images.
```

> We will be downloading the [Chest X-Ray Images (Pneumonia)](https://www.kaggle.com/datasets/paultimothymooney/chest-xray-pneumonia) dataset from [Kaggle](https://www.kaggle.com).

Reviewer 2: If you feel like taking the extra step, you could put this in hf.co/hf-vision hugging face organization as a dataset repository and load from there 😊

Author: Done

You can learn more about the dataset by looking at the Datacard on Kaggle.

> _Data Preprocessing:_

Reviewer 2: ```suggestion
**Data Preprocessing**
```

We will preprocess the lung images by resizing and normalizing the pixel values.

> _DCGAN Architecture:_

Reviewer 2: ```suggestion
**Create the DCGAN Architecture**
```

The generator network takes random noise as input and generates synthetic lung images, while the discriminator network tries to distinguish between real and synthetic images.
Use convolutional layers in both the generator and discriminator to capture spatial features effectively.

We will use the following model architecture:

> Generator(

Reviewer 2: ```suggestion
```bash
Generator(
```
Maybe what's even better is to explain the architecture verbally like in [here](https://github.com/johko/computer-vision-course/pull/57/files#diff-e242ea8730e66737602f9e9bfb772b553da7c9c8846199d2c9552200a3340222R13)

(10): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
(11): ReLU(inplace=True)
(12): ConvTranspose2d(64, 3, kernel_size=(4, 4), stride=(2, 2), padding=(1, 1), bias=False)
(13): Tanh()
)
)

Discriminator(
(main): Sequential(
(0): Conv2d(3, 64, kernel_size=(4, 4), stride=(2, 2), padding=(1, 1), bias=False)
(1): LeakyReLU(negative_slope=0.2, inplace=True)
(2): Conv2d(64, 128, kernel_size=(4, 4), stride=(2, 2), padding=(1, 1), bias=False)
(3): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
(4): LeakyReLU(negative_slope=0.2, inplace=True)
(5): Conv2d(128, 256, kernel_size=(4, 4), stride=(2, 2), padding=(1, 1), bias=False)
(6): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
(7): LeakyReLU(negative_slope=0.2, inplace=True)
(8): Conv2d(256, 512, kernel_size=(4, 4), stride=(2, 2), padding=(1, 1), bias=False)
(9): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
(10): LeakyReLU(negative_slope=0.2, inplace=True)
(11): Conv2d(512, 1, kernel_size=(4, 4), stride=(1, 1), bias=False)
(12): Sigmoid()
)
)

> ## Training:

Reviewer 2: ```suggestion
## Training
```

Train the DCGAN on the prepared lung image dataset.
The generator aims to produce synthetic lung images that are indistinguishable from real images,
while the discriminator learns to correctly classify the images as real or synthetic.
We start by initializing the generator with random noise and will train for 100 epochs.
Let's visualize the progress:

![trainig-gif]()

## How did we do?

We can see that some of the synthetic lung images look good, but some are hazy. Researchers generating synthetic medical images typically

> employee a "human in the middle"--in this case, a expert radiologist--to evaluate the synthetic images. Then, only the images that fool the expert are included with

Reviewer: ```suggestion
employ a "human in the middle"--in this case, an expert radiologist--to evaluate the synthetic images. Then, only the images that fool the expert are included with
```

> with the real data to train the model.

Reviewer: ```suggestion
the real data to train the model.
```

![lung-images]()

> [[Link to Notebook]]

Reviewer: Links are missing.

Given the known challenges associated with employing GANs to train high-quality medical images, some researchers have explored the use of diffusion
models to generate lung images. Medfusion, a conditional latent DDPM for medical images, was proposed in [Diffusion Probabilistic Models Beat GAN On Medical 2D Images](https://arxiv.org/pdf/2212.07501.pdf).
[Synthetically Enhanced: Unveiling Synthetic Data's Potential In Medical Imaging Research](https://arxiv.org/pdf/2311.09402.pdf), Khosravi et. al find that
using a mix of real and synthetically generated lung images using a diffusion process improved model performance.

## Resources and Further Reading

> [A Novel COVID-19 Detection Model Based on DCGAN and Deep Transfer Learning](https://www.sciencedirect.com/science/article/pii/S1877050922007463)

Reviewer 2: Maybe you could put it in list using - for better readability

[Augmentation_Gan](https://github.com/rossettisimone/AUGMENTATION_GAN)
[BrainGAN: Brain MRI Image Generation and Classification Framework Using GAN Architectures and CNN Models](https://www.mdpi.com/1424-8220/22/11/4297)
[Diffusion Probabilistic Models beat GANs on Medical Images](https://arxiv.org/abs/2212.07501)
[DR-DCGAN: A Deep Convolutional Generative Adversarial Network (DC-GAN) for Diabetic Retinopathy Image Synthesis](<https://www.webology.org/data-cms/articles/20220204053948pmwebology%2019%20(2)%20-%2077%20.pdf>)
[Deepfake Image Generation for Improved Brain Tumor Segmentation](https://aps.arxiv.org/abs/2307.14273)
[GANs for Medical Image Synthesis: An Empirical Study](https://arxiv.org/abs/2105.05318)
[Medfusion Github repo](https://github.com/mueller-franzes/medfusion)

> [Medical image editing in the latent space of Generative Adversarial Networks](https://pdf.sciencedirectassets.com/777072/1-s2.0-S2666521221X00026/1-s2.0-S2666521221000168/main.pdf?X-Amz-Security-Token=IQoJb3JpZ2luX2VjEBUaCXVzLWVhc3QtMSJGMEQCIBa2IzG%2BGD7Hjv%2FP20jDKyseswGT785LUQTMmFQjT9ntAiAt%2ByfhQJqJAI2NCOKnVzH%2BxMlTWf2x2uIw2i2OLFOz8yqzBQh%2BEAUaDDA1OTAwMzU0Njg2NSIM9ECuDI1xe%2FlWShlYKpAFeTDTJvita0xvKNmqULElNZxagyOdEwqClxL95y0AJuzQy0JT%2B0LhTDbjmKoo5Ogc5UMO9m2jGU6lz%2BT5Yzt2l90D23PG2J%2BJaJl9ZMJ5KNTYmhfPg4s9gagUsawY7tauMY33wX%2FsgUbWTDQE2Mz7PS0x2YlxGnb%2BebDWUpBiLf%2BoO2%2BGmjA6oV7SltJ1pdF31j3JLngHd12sojW4quFyWSfDBdaj4wCtB%2BVE9CpLBUEnKsHP2DKjzShYc7gCnAYFG1UPNK01u01M4B6%2FAdRi2R9ncY5%2FWybIWZvfncuUv7qbApeKpDbqFvtLb5DovYdVuwp6saKqWWcE%2FGGZdEXioyY%2BLq592cxszrsj72cZ62LKq4mjEt7zoSLPqoVw81BgvsN9VPuBgdlLW0naZA1AD3nYi6Wy%2BUWBjdW6aJ8N5BvY%2B%2BbklAyHIxluEFldW%2BI0sumf7sOdXElgo0j5AumRZdqD9wqDj61YDt%2B86jj6WUkbAbp2kFK2jfaW%2FP6K8yNlCPlGAKee%2FuhOmSPyDidFzwl1du5clnwuIc4CoP3H%2FWLbwKyX1yqXLkvuLeDsIPpjVP8TtOIqLecOWh8RUAntXZ1CNGDI9RZ5ZxT2fBR%2BAih3nYrSnK4G6IQd74QHUi1OFBBjeleDxIkRZcQd6%2BxRV%2FwXvxSMkNSqDiarQxDUSye642M0vDKvDUvdkkIGfh66HwcLabMUyPa3dIRXOVpiPnLKURgrBdEIXsnriMjx2QtEcNDjGZ8S79530EBKvbrOXb3IpDmznAHdIev1wiFenFj1bxsZamn0HSnAUyd7D6d%2Fh%2FB7WeQX5LPFBX2PWKp7aRUaKs7loUT33jcM0PM8gm4aa3qKzPOJdfUDNBAnh0sw45C%2BqwY6sgGhtXuoYqWJBG7SH%2FCkul2joDgNZnUrz0LGR8clMLjdkdIWCb7nc1e0uwXvsNr83g4DPGkfAy%2F1pDKVYAkiTVbtarpBim4E9Thb0nLZTC5wfGQEnanfSSg4qaOaRJqA71bb5E6lYHhWegnp8QCd8YFhYdzr3zu7ob62jm0a1qgVbV1vVLQb05KtZKmFALk0mHVtr9TY7TRO4LuhhAaudJ5NAMejYX5hDniK6gLciipYntn9&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Date=20231205T213802Z&X-Amz-SignedHeaders=host&X-Amz-Expires=300&X-Amz-Credential=ASIAQ3PHCVTYW2PFS26D%2F20231205%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Signature=8977d4b5fb2b4de6557b6ea514c4dd5619738cb8b40bd4b065a0fd0336ba0c26&hash=8db4b852cee5cfe8a167405a15a1356e6963505c298b057892e52f08718b3a03&host=68042c943591013ac2b2430a89b270f6af2c76d8dfd086a07176afe7c76c2c61&pii=S2666521221000168&tid=spdf-617c5bac-a956-4354-813c-b5a5dd6f023c&sid=76355fc83e9a45429678a98306d31967ba09gxrqa&type=client&tsoh=d3d3LnNjaWVuY2VkaXJlY3QuY29t&ua=0f17565253015501520e&rr=830f6f0a4ca66fd4&cc=us)

Reviewer: It looks like the link can be much shorter: 
`https://www.sciencedirect.com/science/article/pii/S2666521221000168?ref=pdf_download&fr=RR-2&rr=833e48fa5e777142`
